usr/src/uts/armv8/os/gicv3.c

```diff
@@ -0,0 +1,1188 @@
+ * redistributors within this region by reading the TYPER register at the
+ * head of each block and using that to advance a pointer through the space
+ * until the GICR_TYPER.Last bit is set, indicating that we've hit the last
+ * redistributor in a region.
+ *
+ * This structure represents the device arena mapping of the redistributor
+ * regions.
+ */
+typedef struct {
+	void		*base;
+	uint64_t	size;
+} gicv3_redist_region_t;
+
+/*
+ * The redistributor structure describes a redistributor in terms of the
+ * redistributor frame addresses in the kernel address space.
+ *
+ * The structure also contains a lock, which must be held when accessing the
+ * redistributor frames, a cached copy of the redistributor type register and
+ * a template to be used when sending software generated interrupts to the
+ * processor associated with this redistributor.
+ *
+ * Redistributor structures are stored in an array, indexed by CPU ID.
+ */
+typedef struct {
+	lock_t			gr_lock;
+	void			*gr_rd_base;
+	void			*gr_sgi_base;
+	void			*gr_vlpi_base;
+	uint64_t		gr_typer;
+	uint64_t		gr_sgir;
+} gicv3_redistributor_t;
+
+typedef struct {
+	/* Base address of the CPU interface */
+	void			*gc_gicc;
+	/* Base address of the distributor */
+	void			*gc_gicd;
+	/* Size of the distributor mapping */
+	uint64_t		gc_gicd_size;
+	/* Shadow copy of the distributor type register */
+	uint32_t		gc_gicd_typer;
+	/* Number of interrupt sources in the traditional interrupt space */
+	uint32_t		gc_maxsources;
+
+	/* Owned mappings of the redistributor regions. */
+	gicv3_redist_region_t	*gc_redist_regions;
+	/* Number of redistributor regions */
+	uint32_t		gc_num_redist_regions;
+	/* Redistributor iteration stride, 0 if no padding pages are present */
+	uint64_t		gc_redist_stride;
+
+	/*
+	 * Redistributors, indexed by CPU ID.
+	 *
+	 * The pointers in these structures index into the mappings owned by
+	 * gc_redist_regions.
+	 */
+	gicv3_redistributor_t	*gc_redist;
+	/* Number of redistributors in the gc_redist array */
+	uint32_t		gc_num_redist;
+
+	/* A flag indicating that we have 32 (or more) priority levels */
+	uint32_t		gc_pri32;
+	/* Protect access the distributor */
+	lock_t		gc_dist_lock;
+	/*
+	 * CPUs for which we have initialized the GIC.  Used to limit IPIs to
+	 * only those CPUs we can target.
+	 */
+	cpuset_t	gc_cpuset;
+} gicv3_conf_t;
+
+static gicv3_conf_t	conf;
+
+#define	GICR_FRAME_SIZE		(64 * 1024)
+
+#define	GIC_IPL_TO_PRI(ipl)	(conf.gc_pri32 ? (GIC_IPL_TO_PRIO((ipl))) : \
+				(GIC_IPL_TO_PRIO16((ipl))))
+
+#define	GICD_LOCK_INIT_HELD()	uint64_t __s = disable_interrupts(); \
+				LOCK_INIT_HELD(&conf.gc_dist_lock)
+#define	GICD_LOCK()		uint64_t __s = disable_interrupts(); \
+				lock_set(&conf.gc_dist_lock)
+#define	GICD_UNLOCK()		lock_clear(&conf.gc_dist_lock); \
+				restore_interrupts(__s)
+
+static inline uint32_t
+reg_read_raw4(void *base, uint32_t reg)
+{
+	return (i_ddi_get32(NULL, (uint32_t *)(base + reg)));
```

Comment:
I think I asked this before, but I don't remember the answer (Sorry).  Is there a reason we're using the `i_ddi_...` implementation, and not the just the actual `ddi_...` public interface?

Reply:
If we pass a NULL `hdlp` to the normal `ddi_*` functions they'll hit a`NULL deference in invoking the function pointer stashed in the handle (which is just `i_ddi_*` in this case).

This is something I'd like to figure out in the future, because this is a bit ugly.